functions/common.py

```diff
@@ -1,26 +1,122 @@
 import exceptions.exceptions as exceptions
 
 
-def factorial(num: int) -> int:
-    if not isinstance(num, int):
-        # Input should be an integer
-        raise exceptions.InputError(num, "Error: non-integer factorial")
-    elif num < 0:
-        # Input should be non-negative
-        raise exceptions.InputError(num, "Error: negative factorial")
-    elif num == 0:
-        return 1
+def factorial(argument: int) -> int:
+    """Returns the product of all non-negative integers less than or equal to some non-negative integer.
+
+    Args:
+        argument (int): Non-negative integer used as input to factorial function
+
+    Returns:
+        int: Product of all non-negative integers less than or equal to 'argument'
+        
+    Raises:
+        InputError: If 'argument' is not a non-negative integer
+    """
+    # Ensure that input is an integer.
+    if int(argument) != argument:
+        raise exceptions.InputError(argument, "Input to factorial should be a non-negative integer.")
+    argument = int(argument)
+    # Ensure that input is non-negative.
+    if argument < 0:
+        raise exceptions.InputError(argument, "Input to factorial should be a non-negative integer.")
+        
+    # Calculate factorial using repeated multiplication.
     result = 1
-    for i in range(1, num + 1):
+    for i in range(1, argument + 1):
         result *= i
     return result
 
 
-def mod(dividend: float, divisor: float) -> float:
-    # TODO not sure on the typing for this,
-    # and should check if can be optimized
-    while dividend > divisor:
-        dividend -= divisor
-    while dividend < divisor:
-        dividend += divisor
-    return dividend
+def is_even(argument: int) -> bool:
+    """Returns true is input is even and false otherwise.
+
+    Args:
+        argument (int): Value to be checked
+
+    Returns:
+        bool: True if 'argument' is even and false otherwise
+        
+    Raises:
+        InputError: If 'argument' is not an integer
+    """
+    # Ensure that argument is an integer.
+    if not isinstance(argument, int):
+        raise exceptions.InputError(exponent, "Parity of non-integers is undefined.")
+        
+    return not (argument & 1)
+
+
+def is_odd(argument: int) -> bool:
```

Comment:
just wondering why `is_odd` isn't just `not is_even` (or vice-versa)? isn't there some duplication here

Reply:
To be honest, I don't remember the logic here. I think I implemented both just to be thorough. I probably figured that is_even is easier to read than not is_odd. Typically, I wouldn't implement either. Instead, I would just write the code in line.

Reply:
I meant more stuff like type checking is duplicated when its exactly the same in both cases, and that the body of `is_even` could just be `return not is_odd` or vice-versa. Don't have strong opinions here though so its non-blocking

Reply:
I meant more stuff like type checking is duplicated when its exactly the same in both cases, and that the body of `is_even` could just be `return not is_odd` or vice-versa. Don't have strong opinions here though so its non-blocking

Reply:
I don't type checking is enforced through definition params iirc. 
Since the def of is_odd will never change, redundancy isn't inherently problematic. 
It just boils down to preference.

Reply:
I think he means the actual type checking being done in the function body. I kept the functions separate because they're pretty simple and I don't want to add extra functions calls. Especially on a small function like this. Truth be told, I could have done it I just prefer not to for the sake of performance (not that we're having any performance issues though).